### ISSUE TITLE
WireGuard: server and client connection are mixed up / reversed

### ISSUE THREAD

**Issue by Prinzhorn** [mitmproxy/mitmproxy#5707]
#### Problem Description

flow.client_conn is the server and flow.server_conn is the client, at least partially

#### Steps to reproduce the behavior:

addon.py

```py
def request(flow):
    if flow.request.pretty_host == "example.com":
        print("flow.client_conn.peername:")
        print(flow.client_conn.peername)

        print("flow.client_conn.sockname:")
        print(flow.client_conn.sockname)


def response(flow):
    if flow.request.pretty_host == "example.com":
        print("flow.server_conn.peername:")
        print(flow.server_conn.peername)

        print("flow.server_conn.sockname:")
        print(flow.server_conn.sockname)

        print("flow.server_conn.address:")
        print(flow.server_conn.address)

```

1. `mitmdump --mode wireguard --scripts addon.py`
2. Load `http://example.com` on my Android through WireGuard
3. `client.sockname` will be `93.184.216.34`, which is the IP of example.com

```
flow.client_conn.peername:
('10.0.0.1', 59929)
flow.client_conn.sockname:
('93.184.216.34', 80) <=================== This is example.com

flow.server_conn.peername:
('93.184.216.34', 80)
flow.server_conn.sockname:
('192.168.178.21', 41144) <=================== This is my local ip
flow.server_conn.address:
('93.184.216.34', 80)

```

#### System Information

```
Mitmproxy: 10.0.0.dev binary
Python:    3.11.0
OpenSSL:   OpenSSL 3.0.5 5 Jul 2022
Platform:  Linux-5.15.0-52-generic-x86_64-with-glibc2.35
```

**Comment by mhils** [comment]
This is _somewhat_ intentional at the moment, at least to the point that it's consistent with transparent mode. 

- The server connections attributes are correct here as far as I can see (please correct me if I'm wrong).
- The client connection peername is correct, but there's some debate to be had about the client sockname. Is that supposed to be the local address we're listening on, or is that supposed to be the address the client is thinking it is connected to (current implementation)? In the former case we need to change the implementation, in the latter case we need to update the docs.

I think there's a point to be made that setting the client sockname to the address we're actually listening on (and not the target host) makes more sense, but I'm happy to be corrected. In either case there shouldn't be much of an information loss. `flow.client_conn.proxy_mode` should be preferred to detect the current server instance, and the destination address is always available in `flow.server_conn.address`.

**Comment by Prinzhorn** [comment]
I have never used transparent mode, so I can't tell if it aligns with it. I was mostly quoting the docs and the IP of example.com is definitely not "The local address we received this connection on.". Unless it is in some twisted way? But like you said this is probably a documentation issue and a problem with the person operating mitmproxy.

> there shouldn't be much of an information loss

The IP of the actual connected device (my Android) is nowhere to be found :thinking: . Like I said, I have no idea how transparent mode is supposed to work, so I can't really make sense of `10.0.0.1`. Is this a feature? If I have multiple devices in a lab, am I responsible to somehow figure out which of them caused a certain flow?

> `flow.client_conn.proxy_mode` should be preferred to detect the current server instance

Yes, I switched to that the second I opened this issue :+1:

**Comment by mhils** [comment]
> I have never used transparent mode, so I can't tell if it aligns with it. I was mostly quoting the docs and the IP of example.com is definitely not "The local address we received this connection on.". Unless it is in some twisted way?

_Technically_ we accept connections for all IPs in transparent mode, which is where this stems from. Right now it's basically aligns with what the client is thinking (reversed of course). But yeah, it's a bit weird.

> Like I said, I have no idea how transparent mode is supposed to work, so I can't really make sense of `10.0.0.1`.

That's your client device's local IP within the WireGuard tunnel. What would you have expected here? the client's IP in your local network?

**Comment by Prinzhorn** [comment]
> That's your client device's local IP within the WireGuard tunnel.

I understand now that this is what this IP technically is. But as a user I don't care about what's _within_ the tunnel. mitmproxy is (ab)using WireGuard/VPN as a tool to capture traffic. As a user I don't care about this, or is this just me? Maybe you can clarify why I should care :smile: 

> What would you have expected here? the client's IP in your local network?

As a user that's the information that has value to me, so yes. But maybe that's just me and others actually want or need the information from within the tunnel. Again, that's just my mental model. That WireGuard/VPN is a tool to make it easier to capture traffic (and it works amazingly well and simple) but other than that, I don't care at all. To me the the `10.0.0.1/32` range has no meaning at all, it's a technical detail.

If I have two WireGuard modes running, then the first client in both would have the same IP, right? And both don't mean anything to me :shrug: 

So maybe we can have an additional property that always has the IPs form outside the tunnel? I mean the IP of the client that initiated the VPN connection, that surely must be known to the WireGuard server?

**Comment by mhils** [comment]
I think you're making a strong argument from the user-perspective here. @decathorpe, how would you feel about passing the sender IP together with the decapsulated WireGuard packets through the stack? We would then essentially have the client's local IP address listed as peername in mitmproxy, and our WireGuard socket listen address as the sockname.

**Comment by decathorpe** [comment]
Can you clarify what you mean by "sender IP"? Which sender, with respect to which connection?

**Comment by mhils** [comment]
Assume your on 192.168.0.1/24, with your laptop on 192.168.0.42 and your phone on 192.168.0.44. You probably want 192.168.0.44? :)

**Comment by decathorpe** [comment]
So ... basically pass along the `src_addr` argument here?
https://github.com/decathorpe/mitmproxy_wireguard/blob/0.1.17/src/wireguard.rs#L219

And then include it in `TransportEvent::DatagramReceived`, `TransportEvent::ConnectionEstablished`, and `TcpStream`, and expose it on the Python side?

**Comment by mhils** [comment]
Yes, exactly. It's kind of a lot of busywork for something where I don't know _how_ useful it is, but technically that'd be it.

**Comment by decathorpe** [comment]
I can look into implementing that. Should not be too much work (famous last words) ...

**Comment by decathorpe** [comment]
https://github.com/decathorpe/mitmproxy_wireguard/pull/22

This only wires up the the support for TCP, I don't think there's a way to do the same for UDP packets that doesn't involve changing how the UDP callback receives arguments.

**Comment by mhils** [comment]
Thanks @decathorpe! I have thought about this for a bit longer and I'm not so sure anymore what's the correct thing to do.
One potential pitfall is that we use the `client_conn` 4-tuple to identify specific TCP connections. By switching to `src_orig` we may end up with a scenario where connection ids are not unique anymore? That's definitely something we need to avoid.

(I haven't fully made my mind up, just sharing this not-fully-fledged intermediary thought)

**Comment by decathorpe** [comment]
I thought this original source address would only be shown in the UI, but not used for internal logic? I wouldn't switch connection_id matching at all.

**Comment by DaClemens** [comment]
was asked to repost my information here

when testing i get these values
```
connection.get_extra_info('peername')=('10.2.3.1', 39202) <- this is the address of the client/peer inside the wireguard tunnel
connection.get_extra_info('sockname')=('0.0.0.0', 51820) <- this is our listening address
connection.get_extra_info('original_src')=('172.19.0.4', 59236) <- this is the address of the peer where we got the last wireguard packet from
connection.get_extra_info('original_dst')=('10.2.3.4', 443) <- this is the address where this connection was headed to
```

from a proxy point of view, this is a connection originating from 10.2.3.1 going to 10.2.3.4, tunneled from 172.19.0.4 to 0.0.0.0 (in lack of a better "local" address here).....

so i'd probably expect either sockname and original_dst to be swapped or peername & original_src

i have to admit though that while i decently understand the mitmproxy_wireguard behavior, i did not look at the mitmproxy codebase, so i might mix some names up here.

**Comment by mhils** [comment]
This is fixed now with mitmproxy_rs:

```python
# peername / sockname represent the tuple from the pespective of the client
writer.get_extra_info('peername')=('10.0.0.1', 45630)
writer.get_extra_info('sockname')=('18.159.197.225', 443)
# original_src/original_dst represent the outer WireGuard connection.
writer.get_extra_info('original_src')=('192.168.178.89', 57952)
writer.get_extra_info('original_dst')=('0.0.0.0', 51820)
```

I think we still need to expose that consistently in the UI though...
